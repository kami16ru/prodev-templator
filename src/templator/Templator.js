export default class Templator {
  #TEMPLATE_REGEXP = /\{\{(.*?)\}\}/gi;

  constructor(template) {
    this._template = template;
  }

  compile(ctx) {
    return this._compileTemplate(ctx);
  }

  _compileTemplate(ctx) {
    let tmpl = this._template;
    let key = null;
    const regExp = this.#TEMPLATE_REGEXP;

    // Важно делать exec именно через константу, иначе уйдёте в бесконечный цикл
    while ((key = regExp.exec(tmpl))) {
      if (key[1]) {
        const tmplValue = key[1].trim();
        // get — функция, написанная ранее в уроке
        const data = this._getData(ctx, tmplValue);
        tmpl = tmpl.replace(new RegExp(key[0], "gi"), data);
      }
    }

    return tmpl;
  }

  _getData(obj, path, defaultValue) {
    const keys = path.split('.');

    let result = obj;
    for (let key of keys) {
      result = result[key];

      if (result === undefined) {
        return defaultValue;
      }
    }

    return result ?? defaultValue; // "??" — [оператор нулевого слияния](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator) (не поддерживается старыми браузерами, для них нужен полифилл)
  }
}
