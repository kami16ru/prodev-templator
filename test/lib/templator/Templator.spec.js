import Templator from "../../../src/lib/templator/Templator"

describe('Class Templator', () => {
  test('new works', () => {
    expect(new Templator()).toBeInstanceOf(Templator)
  })

  test('_template is set', () => {
    const template = '<div>Hello<div>'
    const tmpl = new Templator(template)

    expect(tmpl._template).toBe(template)
  })
})

describe('compile', () => {
  test('works', () => {
    let testTempl = `
      <div>
          {{ field1 }}
          <span>{{field2}}</span>
          <span>{{ field3.info.name }}</span>
      </div>
    `
    const ctx = {
      field1: 'Text 1',
      field2: 42,
      field3: {
        info: {
          name: 'Simon',
        }
      }
    }
    const tmpl = new Templator(testTempl)

    expect(tmpl.compile(ctx)).toMatchSnapshot()
  })
})

describe('get', () => {
  const obj = {
    user: {
      isAdmin: false,
      isPoet: true,
      info: {
        firstName: 'Alexander',
        lastName: 'Pushkin'
      }
    }
  };
  const tmpl = new Templator()
  const get = tmpl._getData

  test('second level works', () => {
    expect(get(obj, 'user.isPoet')).toBe(true)
  })

  test('third level works', () => {
    expect(get(obj, 'user.info.firstName')).toBe('Alexander')
  })

  test('third level returns undefined if no key', () => {
    expect(get(obj, 'user.info.contacts')).toBe(undefined)
  })

  test('forth level returns undefined if no key', () => {
    expect(get(obj, 'user.info.contacts.email')).toBe(undefined)
  })

  test('returns default value if set', () => {
    expect(get(obj, 'user.info.contacts.email', 'a.pushkin@ya.ru')).toBe('a.pushkin@ya.ru')

  })

  test('cannot mutate if already set', () => {
    expect(get(obj, 'user.isAdmin', true)).toBe(false)
  })
})
