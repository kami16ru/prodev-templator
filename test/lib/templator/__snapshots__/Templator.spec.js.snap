// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compile works 1`] = `
"
      <div>
          Text 1
          <span>42</span>
          <span>Simon</span>
      </div>
    "
`;
